docs/docs/30-administration/10-server-config.md
# Server configuration

## User registration

Registration is closed by default. While disabled an administrator needs to add new users manually (exp. `woodpecker-cli user add`).

If registration is open every user with an account at the configured [forges](./11-forges/10-overview.md) can login to Woodpecker.
This example enables open registration for users that are members of approved organizations:

```diff
# docker-compose.yml
version: '3'

services:
  woodpecker-server:
    [...]
    environment:
      - [...]
+     - WOODPECKER_OPEN=true
+     - WOODPECKER_ORGS=dolores,dogpatch

```

## Administrators

Administrators should also be enumerated in your configuration.

```diff
# docker-compose.yml
version: '3'

services:
  woodpecker-server:
    [...]
    environment:
      - [...]
+     - WOODPECKER_ADMIN=johnsmith,janedoe
```

## Filtering repositories

Woodpecker operates with the user's OAuth permission. Due to the coarse permission handling of GitHub, you may end up syncing more repos into Woodpecker than preferred.

Use the `WOODPECKER_REPO_OWNERS` variable to filter which GitHub user's repos should be synced only. You typically want to put here your company's GitHub name.

```diff
# docker-compose.yml
version: '3'

services:
  woodpecker-server:
    [...]
    environment:
      - [...]
+     - WOODPECKER_REPO_OWNERS=mycompany,mycompanyossgithubuser
```

## Global registry setting

If you want to make available a specific private registry to all pipelines, use the `WOODPECKER_DOCKER_CONFIG` server configuration.
Point it to your server's docker config.

```diff
# docker-compose.yml
version: '3'

services:
  woodpecker-server:
    [...]
    environment:
      - [...]
+     - WOODPECKER_DOCKER_CONFIG=/home/user/.docker/config.json
```

## Handling sensitive data in docker-compose and docker-swarm

To handle sensitive data in docker-compose or docker-swarm configurations there are several options:

For docker-compose you can use a .env file next to your compose configuration to store the secrets outside of the compose file. While this separates configuration from secrets it is still not very secure.

Alternatively use docker-secrets. As it may be difficult to use docker secrets for environment variables woodpecker allows to read sensible data from files by providing a `*_FILE` option of all sensible configuration variables. Woodpecker will try to read the value directly from this file. Keep in mind that when the original environment variable gets specified at the same time it will override the value read from the file.

```diff
# docker-compose.yml
version: '3'

services:
  woodpecker-server:
    [...]
    environment:
      - [...]
+     - WOODPECKER_AGENT_SECRET_FILE=/run/secrets/woodpecker-agent-secret
+   secrets:
+     - woodpecker-agent-secret
+
+secrets:
+  woodpecker-agent-secret:
+    external: true
```

Store a value to a docker secret like this:
`echo "my_agent_secret_key" | docker secret create woodpecker-agent-secret -`

or generate a random one like this:

`openssl rand -hex 32 | docker secret create woodpecker-agent-secret -`

## All server configuration options

The following list describes all available server configuration options.

### `WOODPECKER_LOG_LEVEL`
> Default: empty

Configures the logging level. Possible values are `trace`, `debug`, `info`, `warn`, `error`, `fatal`, `panic`, `disabled` and empty.

### `WOODPECKER_DEBUG_PRETTY`
> Default: `false`

Enable pretty-printed debug output.

### `WOODPECKER_DEBUG_NOCOLOR`
> Default: `true`

Disable colored debug output.

### `WOODPECKER_HOST`
> Default: empty

Server fully qualified URL of the user-facing hostname.

Example: `WOODPECKER_HOST=http://woodpecker.example.org`

### `WOODPECKER_SERVER_ADDR`
> Default: `:8000`

Configures the HTTP listener port.

### `WOODPECKER_SERVER_CERT`
> Default: empty

Path to an SSL certificate used by the server to accept HTTPS requests.

Example: `WOODPECKER_SERVER_CERT=/path/to/cert.pem`

### `WOODPECKER_SERVER_KEY`
> Default: empty

Path to an SSL certificate key used by the server to accept HTTPS requests.

Example: `WOODPECKER_SERVER_KEY=/path/to/key.pem`

### `WOODPECKER_LETS_ENCRYPT`
> Default: `false`

Automatically generates an SSL certificate using Let's Encrypt, and configures the server to accept HTTPS requests.

### `WOODPECKER_GRPC_ADDR`
> Default: `:9000`

Configures the gRPC listener port.


### `WOODPECKER_ADMIN`
> Default: empty

Comma-separated list of admin accounts.

Example: `WOODPECKER_ADMIN=user1,user2`

### `WOODPECKER_ORGS`
> Default: empty

Comma-separated list of approved organizations.

Example: `org1,org2`

### `WOODPECKER_REPO_OWNERS`
> Default: empty

Comma-separated list of syncable repo owners. ???

Example: `user1,user2`

### `WOODPECKER_OPEN`
> Default: `false`

Enable to allow user registration.

### `WOODPECKER_DOCS`
> Default: `https://woodpecker-ci.org/`

Link to documentation in the UI.

### `WOODPECKER_AUTHENTICATE_PUBLIC_REPOS`
> Default: `false`

Always use authentication to clone repositories even if they are public. Needed if the forge requires to always authenticate as used by many companies.

### `WOODPECKER_DEFAULT_CANCEL_PREVIOUS_PIPELINE_EVENTS`
> Default: `pull_request, push`

List of event names that will be canceled when a new pipeline for the same context (tag, branch) is created.

### `WOODPECKER_DEFAULT_CLONE_IMAGE`
> Default: `woodpeckerci/plugin-git:latest`

The default docker image to be used when cloning the repo

### `WOODPECKER_SESSION_EXPIRES`
> Default: `72h`

Configures the session expiration time.

### `WOODPECKER_ESCALATE`
> Default: `plugins/docker,plugins/gcr,plugins/ecr,woodpeckerci/plugin-docker,woodpeckerci/plugin-docker-buildx`

Docker images to run in privileged mode. Only change if you are sure what you do!

<!--
### `WOODPECKER_VOLUME`
> Default: empty

Comma-separated list of Docker volumes that are mounted into every pipeline step.

Example: `WOODPECKER_VOLUME=/path/on/host:/path/in/container:rw`|
-->

### `WOODPECKER_DOCKER_CONFIG`
> Default: empty

Configures a specific private registry config for all pipelines.

Example: `WOODPECKER_DOCKER_CONFIG=/home/user/.docker/config.json`

<!--
### `WOODPECKER_ENVIRONMENT`
> Default: empty

TODO

### `WOODPECKER_NETWORK`
> Default: empty

Comma-separated list of Docker networks that are attached to every pipeline step.

Example: `WOODPECKER_NETWORK=network1,network2`
-->

### `WOODPECKER_AGENT_SECRET`
> Default: empty

A shared secret used by server and agents to authenticate communication. A secret can be generated by `openssl rand -hex 32`.

### `WOODPECKER_AGENT_SECRET_FILE`
> Default: empty

Read the value for `WOODPECKER_AGENT_SECRET` from the specified filepath

### `WOODPECKER_KEEPALIVE_MIN_TIME`
> Default: empty

Server-side enforcement policy on the minimum amount of time a client should wait before sending a keepalive ping.

Example: `WOODPECKER_KEEPALIVE_MIN_TIME=10s`

### `WOODPECKER_DATABASE_DRIVER`
> Default: `sqlite3`

The database driver name. Possible values are `sqlite3`, `mysql` or `postgres`.

### `WOODPECKER_DATABASE_DATASOURCE`
> Default: `woodpecker.sqlite`

The database connection string. The default value is the path of the embedded SQLite database file.

Example:
```bash
# MySQL
# https://github.com/go-sql-driver/mysql#dsn-data-source-name
WOODPECKER_DATABASE_DATASOURCE=root:password@tcp(1.2.3.4:3306)/woodpecker?parseTime=true

# PostgreSQL
# https://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-CONNSTRING
WOODPECKER_DATABASE_DATASOURCE=postgres://root:password@1.2.3.4:5432/woodpecker?sslmode=disable
```

### `WOODPECKER_DATABASE_DATASOURCE_FILE`
> Default: empty

Read the value for `WOODPECKER_DATABASE_DATASOURCE` from the specified filepath

### `WOODPECKER_PROMETHEUS_AUTH_TOKEN`
> Default: empty

Token to secure the Prometheus metrics endpoint.

### `WOODPECKER_PROMETHEUS_AUTH_TOKEN_FILE`
> Default: empty

Read the value for `WOODPECKER_PROMETHEUS_AUTH_TOKEN` from the specified filepath

### `WOODPECKER_STATUS_CONTEXT`
> Default: `ci/woodpecker`

Context prefix Woodpecker will use to publish status messages to SCM. You probably will only need to change it if you run multiple Woodpecker instances for a single repository.

### `WOODPECKER_STATUS_CONTEXT_FORMAT`
> Default: `{{ .context }}/{{ .event }}/{{ .pipeline }}`

Template for the status messages published to forges, uses [Go templates](https://pkg.go.dev/text/template) as template language.
Supported variables:
- `context`: Woodpecker's context (see `WOODPECKER_STATUS_CONTEXT`)
- `event`: the event which started the pipeline
- `pipeline`: the pipeline's name
- `owner`: the repo's owner
- `repo`: the repo's name

---

### `WOODPECKER_LIMIT_MEM_SWAP`
> Default: `0`

The maximum amount of memory a single pipeline container is allowed to swap to disk, configured in bytes. There is no limit if `0`.

### `WOODPECKER_LIMIT_MEM`
> Default: `0`

The maximum amount of memory a single pipeline container can use, configured in bytes. There is no limit if `0`.

### `WOODPECKER_LIMIT_SHM_SIZE`
> Default: `0`

The maximum amount of memory of `/dev/shm` allowed in bytes. There is no limit if `0`.

### `WOODPECKER_LIMIT_CPU_QUOTA`
> Default: `0`

The number of microseconds per CPU period that the container is limited to before throttled. There is no limit if `0`.

### `WOODPECKER_LIMIT_CPU_SHARES`
> Default: `0`

The relative weight vs. other containers.


CHANGELOG.md
# Changelog

## [0.15.5](https://github.com/woodpecker-ci/woodpecker/releases/tag/v0.15.5) - 2022-10-13

* BUGFIXES
  * Change build message column type to text (#1252) (#1253)
* ENHANCEMENTS
  * Bump DefaultCloneImage version to v1.6.0 (#1254)
  * On Repo update, keep old "Clone" if update would empty it (#1170) (#1195)

## [0.15.4](https://github.com/woodpecker-ci/woodpecker/releases/tag/v0.15.4) - 2022-09-06

* BUGFIXES
  * Extract commit message from branch creation (#1150) (#1153)
  * Respect WOODPECKER_GITEA_SKIP_VERIFY (#1152) (#1151)
  * update golang.org/x/crypto (#1124)
  * Implement Refresher for GitLab (#1031) (#1120)
  * Make returned proc list to be returned always in correct order (#1060) (#1065)
  * Update type of 'log_data' from blob to longblob (#1050) (#1052)
  * Make ListItem component more accessible by using a button tag when clickable (#1044) (#1046)
* MISC
  * Update base images (#1024) (#1025)

## [0.15.3](https://github.com/woodpecker-ci/woodpecker/releases/tag/v0.15.3) - 2022-06-16

* SECURITY
  * Update github.com/containerd/containerd (#978) (#980)
* BUGFIXES
  * Return to page after clicking login at navbar (#975) (#976)

## [0.15.2](https://github.com/woodpecker-ci/woodpecker/releases/tag/v0.15.2) - 2022-06-14

* BUGFIXES
  * Fix uppercase from_secrets (#842) (#925)
  * Fix key/val format for dind env vars (#889) (#890)
  * Update helm chart releasing (#882) (#888)
* DOCUMENTATION
  * Fix run_on references with runs_on in docs (#965)

## [0.15.1](https://github.com/woodpecker-ci/woodpecker/releases/tag/v0.15.1) - 2022-04-13

* SECURITY
  * Escape html / xml in log view (#879) (#880)
* FEATURES
  * Build multiarch images for server (#821) (#822)
* BUGFIXES
  * Branch list enhancements (#808) (#809)
  * Get Netrc machine from clone url (#800) (#803)

## [v0.15.0](https://github.com/woodpecker-ci/woodpecker/releases/tag/v0.15.0) - 2022-02-24

* BREAKING
  * Change paths to use woodpecker instead of drone (#494)
  * Move plugin config to root.pipeline.[step].settings (#464)
  * Replace debug with log-level flag (#440)
  * Change prometheus metrics from `drone_*` to `woodpecker_*` (#439)
  * Replace DRONE_ with CI_ variables in pipeline steps (#427)
  * Enable pull_request hook by default on repository activation (#420)
  * Remote Gitea drop basic auth support (#365)
  * Change pipeline config path resolution (#299)
  * Remove push, tag and deployment webhook filters (#281)
  * Clean up config environment variables for server and agent (#218)
* SECURITY
  * Add linter bidichk to prevent malicious utf8 chars (#516)
* FEATURES
  * Show changed files of pipeline in UI (#650)
  * Show yml config of pipeline in UI (#649)
  * Multiarch build for cli and agent docker images (#634), (#622)
  * Get secrets in settings (#604)
  * Add multi-pipeline support to exec & lint (#568)
  * Add repo branches endpoint (#481)
  * Add repo permission endpoint (#436)
  * Add web-config endpoint (#433)
  * Replace www-path with www-proxy option for development (#248)
* BUGFIXES
  * Make gRPC error "too many keepalive pings" only show up in trace logs (#787)
  * WOODPECKER_ENVIRONMENT: ignore items only containing a key and no value (#781)
  * Fix pipeline timestamps (#730)
  * Remove "panic()" as much as possible from code (#682)
  * Send decline events back to UI (#680)
  * Notice all changed files of all related commits for gitea push webhooks (#675)
  * Use global branch filter only on events containing branch info (#659)
  * API GetRepos() return empty list if no active repos exist (#658)
  * Skip nested GitLab repositories during sync (#656), (#652)
  * Build proc tree function should not depend on sorted procs list (#647)
  * Fix sqlite migration on column drop of abnormal schemas (#629)
  * Fix gRPC incompatibility in helm chart (#627)
  * Fix new pipeline not published to UI if protected repo mode enabled (#619)
  * Dont panic, report error back (#582)
  * Improve status updates (#561)
  * Let normal repo admins change timeout to lower values (#543)
  * Fix registry delete (#532)
  * Fix overflowing commit messages (#528)
  * Fix passing of netrc credentials to clone step (#492)
  * Fix various typos (#416)
  * Append trailing slash to default GH API URL (#411)
  * Fix filter pipeline config files (#279)
* ENHANCEMENTS
  * Return better error if repo was deleted/renamed (#780)
  * Add support to set default clone image via environment variable (#769)
  * Add flag to always authenticate when cloning public repositories from locked down / private only forges (#760)
  * UI: show date time on hover over time items (#756)
  * Add repo-link to badge markdown in UI (#753)

cmd/server/flags.go
// Copyright 2019 Laszlo Fogas
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"os"
	"time"

	"github.com/urfave/cli/v2"

	"github.com/woodpecker-ci/woodpecker/shared/constant"
)

var flags = []cli.Flag{
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_LOG_LEVEL"},
		Name:    "log-level",
		Usage:   "set logging level",
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_DEBUG_PRETTY"},
		Name:    "pretty",
		Usage:   "enable pretty-printed debug output",
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_DEBUG_NOCOLOR"},
		Name:    "nocolor",
		Usage:   "disable colored debug output",
		Value:   true,
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_HOST"},
		Name:    "server-host",
		Usage:   "server fully qualified url (<scheme>://<host>)",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_SERVER_ADDR"},
		Name:    "server-addr",
		Usage:   "server address",
		Value:   ":8000",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_SERVER_CERT"},
		Name:    "server-cert",
		Usage:   "server ssl cert path",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_SERVER_KEY"},
		Name:    "server-key",
		Usage:   "server ssl key path",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_LETS_ENCRYPT_EMAIL"},
		Name:    "lets-encrypt-email",
		Usage:   "let's encrypt email",
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_LETS_ENCRYPT"},
		Name:    "lets-encrypt",
		Usage:   "enable let's encrypt",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_GRPC_ADDR"},
		Name:    "grpc-addr",
		Usage:   "grpc address",
		Value:   ":9000",
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_ADMIN"},
		Name:    "admin",
		Usage:   "list of admin users",
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_ORGS"},
		Name:    "orgs",
		Usage:   "list of approved organizations",
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_REPO_OWNERS"},
		Name:    "repo-owners",
		Usage:   "List of syncable repo owners",
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_OPEN"},
		Name:    "open",
		Usage:   "enable open user registration",
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_AUTHENTICATE_PUBLIC_REPOS"},
		Name:    "authenticate-public-repos",
		Usage:   "Always use authentication to clone repositories even if they are public. Needed if the SCM requires to always authenticate as used by many companies.",
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_DEFAULT_CANCEL_PREVIOUS_PIPELINE_EVENTS"},
		Name:    "default-cancel-previous-pipeline-events",
		Usage:   "List of event names that will be canceled when a new pipeline for the same context (tag, branch) is created.",
		Value:   cli.NewStringSlice("push", "pull_request"),
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_DEFAULT_CLONE_IMAGE"},
		Name:    "default-clone-image",
		Usage:   "The default docker image to be used when cloning the repo",
		Value:   constant.DefaultCloneImage,
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_DOCS"},
		Name:    "docs",
		Usage:   "link to user documentation",
		Value:   "https://woodpecker-ci.org/",
	},
	&cli.DurationFlag{
		EnvVars: []string{"WOODPECKER_SESSION_EXPIRES"},
		Name:    "session-expires",
		Usage:   "session expiration time",
		Value:   time.Hour * 72,
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_ESCALATE"},
		Name:    "escalate",
		Usage:   "images to run in privileged mode",
		Value:   cli.NewStringSlice(constant.PrivilegedPlugins...),
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_VOLUME"},
		Name:    "volume",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_DOCKER_CONFIG"},
		Name:    "docker-config",
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_ENVIRONMENT"},
		Name:    "environment",
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_NETWORK"},
		Name:    "network",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_AGENT_SECRET"},
		Name:     "agent-secret",
		Usage:    "server-agent shared password",
		FilePath: os.Getenv("WOODPECKER_AGENT_SECRET_FILE"),
	},
	&cli.DurationFlag{
		EnvVars: []string{"WOODPECKER_KEEPALIVE_MIN_TIME"},
		Name:    "keepalive-min-time",
		Usage:   "server-side enforcement policy on the minimum amount of time a client should wait before sending a keepalive ping.",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_SECRET_ENDPOINT"},
		Name:    "secret-service",
		Usage:   "secret plugin endpoint",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_REGISTRY_ENDPOINT"},
		Name:    "registry-service",
		Usage:   "registry plugin endpoint",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_CONFIG_SERVICE_ENDPOINT"},
		Name:    "config-service-endpoint",
		Usage:   "url used for calling configuration service endpoint",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_DATABASE_DRIVER"},
		Name:    "driver",
		Usage:   "database driver",
		Value:   "sqlite3",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_DATABASE_DATASOURCE"},
		Name:     "datasource",
		Usage:    "database driver configuration string",
		Value:    "woodpecker.sqlite",
		FilePath: os.Getenv("WOODPECKER_DATABASE_DATASOURCE_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_PROMETHEUS_AUTH_TOKEN"},
		Name:     "prometheus-auth-token",
		Usage:    "token to secure prometheus metrics endpoint",
		Value:    "",
		FilePath: os.Getenv("WOODPECKER_PROMETHEUS_AUTH_TOKEN_FILE"),
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_STATUS_CONTEXT", "WOODPECKER_GITHUB_CONTEXT", "WOODPECKER_GITEA_CONTEXT"},
		Name:    "status-context",
		Usage:   "status context prefix",
		Value:   "ci/woodpecker",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_STATUS_CONTEXT_FORMAT"},
		Name:    "status-context-format",
		Usage:   "status context format",
		Value:   "{{ .context }}/{{ .event }}/{{ .pipeline }}",
	},
	//
	// resource limit parameters
	//
	&cli.Int64Flag{
		EnvVars: []string{"WOODPECKER_LIMIT_MEM_SWAP"},
		Name:    "limit-mem-swap",
		Usage:   "maximum swappable memory allowed in bytes",
	},
	&cli.Int64Flag{
		EnvVars: []string{"WOODPECKER_LIMIT_MEM"},
		Name:    "limit-mem",
		Usage:   "maximum memory allowed in bytes",
	},
	&cli.Int64Flag{
		EnvVars: []string{"WOODPECKER_LIMIT_SHM_SIZE"},
		Name:    "limit-shm-size",
		Usage:   "docker compose /dev/shm allowed in bytes",
	},
	&cli.Int64Flag{
		EnvVars: []string{"WOODPECKER_LIMIT_CPU_QUOTA"},
		Name:    "limit-cpu-quota",
		Usage:   "impose a cpu quota",
	},
	&cli.Int64Flag{
		EnvVars: []string{"WOODPECKER_LIMIT_CPU_SHARES"},
		Name:    "limit-cpu-shares",
		Usage:   "change the cpu shares",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_LIMIT_CPU_SET"},
		Name:    "limit-cpu-set",
		Usage:   "set the cpus allowed to execute containers",
	},
	//
	// GitHub
	//
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GITHUB"},
		Name:    "github",
		Usage:   "github driver is enabled",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_GITHUB_URL"},
		Name:    "github-server",
		Usage:   "github server address",
		Value:   "https://github.com",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_GITHUB_CLIENT"},
		Name:     "github-client",
		Usage:    "github oauth2 client id",
		FilePath: os.Getenv("WOODPECKER_GITHUB_CLIENT_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_GITHUB_SECRET"},
		Name:     "github-secret",
		Usage:    "github oauth2 client secret",
		FilePath: os.Getenv("WOODPECKER_GITHUB_SECRET_FILE"),
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GITHUB_MERGE_REF"},
		Name:    "github-merge-ref",
		Usage:   "github pull requests use merge ref",
		Value:   true,
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GITHUB_SKIP_VERIFY"},
		Name:    "github-skip-verify",
		Usage:   "github skip ssl verification",
	},
	//
	// Gogs
	//
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GOGS"},
		Name:    "gogs",
		Usage:   "gogs driver is enabled",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_GOGS_URL"},
		Name:    "gogs-server",
		Usage:   "gogs server address",
		Value:   "https://try.gogs.io",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_GOGS_GIT_USERNAME"},
		Name:     "gogs-git-username",
		Usage:    "gogs service account username",
		FilePath: os.Getenv("WOODPECKER_GOGS_GIT_USERNAME_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_GOGS_GIT_PASSWORD"},
		Name:     "gogs-git-password",
		Usage:    "gogs service account password",
		FilePath: os.Getenv("WOODPECKER_GOGS_GIT_PASSWORD_FILE"),
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GOGS_PRIVATE_MODE"},
		Name:    "gogs-private-mode",
		Usage:   "gogs private mode enabled",
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GOGS_SKIP_VERIFY"},
		Name:    "gogs-skip-verify",
		Usage:   "gogs skip ssl verification",
	},
	//
	// Gitea
	//
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GITEA"},
		Name:    "gitea",
		Usage:   "gitea driver is enabled",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_GITEA_URL"},
		Name:    "gitea-server",
		Usage:   "gitea server address",
		Value:   "https://try.gitea.io",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_GITEA_CLIENT"},
		Name:     "gitea-client",
		Usage:    "gitea oauth2 client id",
		FilePath: os.Getenv("WOODPECKER_GITEA_CLIENT_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_GITEA_SECRET"},
		Name:     "gitea-secret",
		Usage:    "gitea oauth2 client secret",
		FilePath: os.Getenv("WOODPECKER_GITEA_SECRET_FILE"),
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GITEA_SKIP_VERIFY"},
		Name:    "gitea-skip-verify",
		Usage:   "gitea skip ssl verification",
	},
	//
	// Bitbucket
	//
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_BITBUCKET"},
		Name:    "bitbucket",
		Usage:   "bitbucket driver is enabled",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_BITBUCKET_CLIENT"},
		Name:     "bitbucket-client",
		Usage:    "bitbucket oauth2 client id",
		FilePath: os.Getenv("WOODPECKER_BITBUCKET_CLIENT_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_BITBUCKET_SECRET"},
		Name:     "bitbucket-secret",
		Usage:    "bitbucket oauth2 client secret",
		FilePath: os.Getenv("WOODPECKER_BITBUCKET_SECRET_FILE"),
	},
	//
	// GitLab
	//
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GITLAB"},
		Name:    "gitlab",
		Usage:   "gitlab driver is enabled",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_GITLAB_URL"},
		Name:    "gitlab-server",
		Usage:   "gitlab server address",
		Value:   "https://gitlab.com",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_GITLAB_CLIENT"},
		Name:     "gitlab-client",
		Usage:    "gitlab oauth2 client id",
		FilePath: os.Getenv("WOODPECKER_GITLAB_CLIENT_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_GITLAB_SECRET"},
		Name:     "gitlab-secret",
		Usage:    "gitlab oauth2 client secret",
		FilePath: os.Getenv("WOODPECKER_GITLAB_SECRET_FILE"),
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_GITLAB_SKIP_VERIFY"},
		Name:    "gitlab-skip-verify",
		Usage:   "gitlab skip ssl verification",
	},
	//
	// Bitbucket Stash
	//
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_STASH"},
		Name:    "stash",
		Usage:   "stash driver is enabled",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_STASH_URL"},
		Name:    "stash-server",
		Usage:   "stash server address",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_STASH_CONSUMER_KEY"},
		Name:     "stash-consumer-key",
		Usage:    "stash oauth1 consumer key",
		FilePath: os.Getenv("WOODPECKER_STASH_CONSUMER_KEY_FILE"),
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_STASH_CONSUMER_RSA"},
		Name:    "stash-consumer-rsa",
		Usage:   "stash oauth1 private key file",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_STASH_CONSUMER_RSA_STRING"},
		Name:    "stash-consumer-rsa-string",
		Usage:   "stash oauth1 private key string",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_STASH_GIT_USERNAME"},
		Name:     "stash-git-username",
		Usage:    "stash service account username",
		FilePath: os.Getenv("WOODPECKER_STASH_GIT_USERNAME_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_STASH_GIT_PASSWORD"},
		Name:     "stash-git-password",
		Usage:    "stash service account password",
		FilePath: os.Getenv("WOODPECKER_STASH_GIT_PASSWORD_FILE"),
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_STASH_SKIP_VERIFY"},
		Name:    "stash-skip-verify",
		Usage:   "stash skip ssl verification",
	},
	//
	// Coding
	//
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_CODING"},
		Name:    "coding",
		Usage:   "coding driver is enabled",
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_CODING_URL"},
		Name:    "coding-server",
		Usage:   "coding server address",
		Value:   "https://coding.net",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_CODING_CLIENT"},
		Name:     "coding-client",
		Usage:    "coding oauth2 client id",
		FilePath: os.Getenv("WOODPECKER_CODING_CLIENT_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_CODING_SECRET"},
		Name:     "coding-secret",
		Usage:    "coding oauth2 client secret",
		FilePath: os.Getenv("WOODPECKER_CODING_SECRET_FILE"),
	},
	&cli.StringSliceFlag{
		EnvVars: []string{"WOODPECKER_CODING_SCOPE"},
		Name:    "coding-scope",
		Usage:   "coding oauth scope",
		Value: cli.NewStringSlice(
			"user",
			"project",
			"project:depot",
		),
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_CODING_GIT_MACHINE"},
		Name:    "coding-git-machine",
		Usage:   "coding machine name",
		Value:   "git.coding.net",
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_CODING_GIT_USERNAME"},
		Name:     "coding-git-username",
		Usage:    "coding machine user username",
		FilePath: os.Getenv("WOODPECKER_CODING_GIT_USERNAME_FILE"),
	},
	&cli.StringFlag{
		EnvVars:  []string{"WOODPECKER_CODING_GIT_PASSWORD"},
		Name:     "coding-git-password",
		Usage:    "coding machine user password",
		FilePath: os.Getenv("WOODPECKER_CODING_GIT_PASSWORD_FILE"),
	},
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_CODING_SKIP_VERIFY"},
		Name:    "coding-skip-verify",
		Usage:   "coding skip ssl verification",
	},
	//
	// development flags
	//
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_DEV_WWW_PROXY"},
		Name:    "www-proxy",
		Usage:   "serve the website by using a proxy (used for development)",
		Hidden:  true,
	},
	&cli.StringFlag{
		EnvVars: []string{"WOODPECKER_DEV_OAUTH_HOST"},
		Name:    "server-dev-oauth-host",
		Usage:   "server fully qualified url (<scheme>://<host>) used for oauth redirect (used for development)",
		Value:   "",
		Hidden:  true,
	},
	//
	// misc
	//
	&cli.BoolFlag{
		EnvVars: []string{"WOODPECKER_FLAT_PERMISSIONS"},
		Name:    "flat-permissions",
		Usage:   "no forge call for permissions should be made",
		Hidden:  true,
		// TODO(485) temporary workaround to not hit api rate limits
	},
}
